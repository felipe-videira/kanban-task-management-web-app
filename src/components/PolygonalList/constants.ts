export const SCALE_ROTATION_45DEG = 1.4142;
export const POINTING_UP_ROTATION = 45;
export const POINTING_DOWN_ROTATION = 225;
