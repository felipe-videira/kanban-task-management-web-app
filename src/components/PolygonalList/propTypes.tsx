import { elementType, number, object, arrayOf, bool } from "prop-types";

export default {
  data: arrayOf(object).isRequired,
  ItemComponent: elementType.isRequired,
  itemProps: object,
  itemSize: number.isRequired,
  pointingUp: bool,
};
