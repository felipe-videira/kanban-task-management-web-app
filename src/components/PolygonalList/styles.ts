import styled from "styled-components/macro";
import {
  POINTING_DOWN_ROTATION,
  POINTING_UP_ROTATION,
  SCALE_ROTATION_45DEG,
} from "./constants";

type ContainerProps = {
  readonly size: number;
  readonly pointingUp?: boolean | null;
};

type ItemContainerProps = {
  readonly length: number;
  readonly itemSize: number;
  readonly zIndex: number;
  readonly polygonExteriorAngle: number;
};

type ItemContentProps = {
  readonly index: number;
  readonly pointingUp?: boolean | null;
} & ItemContainerProps;

export const PolygonalListContainer = styled.div.attrs<ContainerProps>(
  (props) => ({
    size: `${props.size / SCALE_ROTATION_45DEG}px`,
  })
)<ContainerProps>`
  ${(props) => `
    transform: rotate(${
      props.pointingUp ? POINTING_UP_ROTATION : POINTING_DOWN_ROTATION
    }deg);
    width: ${props.size};
    height: ${props.size};
`}
`;

export const PolygonalListItemContainer = styled.div.attrs<ItemContainerProps>(
  (props) => ({
    style: {
      zIndex: props.zIndex,
      transform: `rotate(${props.polygonExteriorAngle}deg)`,
    },
  })
)<ItemContainerProps>`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;

  ${(props) => {
    if (props.length >= 3) {
      const origin = `-${props.itemSize / (props.itemSize * 0.1)}px`;

      return `
        &::after {
            content: "";
            position: absolute;
            transform-origin: ${origin} ${origin};
            height: calc(100% / (${props.length}/ 4));
            width: ${props.itemSize * 0.1}px;
            background: rgba(0,0,0,0.2);
            top: 0;
            left: 0;
            transform: rotate(${(90 - props.polygonExteriorAngle) / 2}deg);
        }
      `;
    }

    return "";
  }}
`;

export const PolygonalListItemContent = styled.div.attrs<ItemContentProps>(
  (props) => {
    const width = `${props.itemSize}px`;
    const top = `-${props.itemSize / 2}px`;
    const rotation =
      props.polygonExteriorAngle * (props.index + 1) +
      (props.pointingUp ? POINTING_UP_ROTATION : POINTING_DOWN_ROTATION);

    return {
      style: {
        zIndex: props.zIndex,
        top,
        left: top,
        width,
        height: width,
        transform: `rotate(-${rotation}deg)`,
      },
    };
  }
)<ItemContentProps>`
  position: absolute;
`;
