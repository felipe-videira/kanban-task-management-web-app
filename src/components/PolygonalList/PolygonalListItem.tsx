import { InferProps, number } from "prop-types";
import {
  PolygonalListItemContainer as ItemContainer,
  PolygonalListItemContent as ItemContent,
} from "./styles";
import propTypes from "./propTypes";

function PolygonalListItem(
  props: InferProps<typeof PolygonalListItem.propTypes>
) {
  const zIndex = props.length - props.index;

  return (
    <ItemContainer
      zIndex={zIndex}
      length={props.length}
      itemSize={props.itemSize}
      polygonExteriorAngle={props.polygonExteriorAngle}
    >
      <ItemContent
        zIndex={zIndex}
        index={props.index}
        length={props.length}
        itemSize={props.itemSize}
        pointingUp={props.pointingUp}
        polygonExteriorAngle={props.polygonExteriorAngle}
      >
        <props.ItemComponent
          {...props.data[props.index]}
          {...props.itemProps}
        />
      </ItemContent>

      {props.index < props.length - 1 && (
        <PolygonalListItem {...props} index={props.index + 1} />
      )}
    </ItemContainer>
  );
}

PolygonalListItem.propTypes = {
  ...propTypes,
  index: number.isRequired,
  length: number.isRequired,
  polygonExteriorAngle: number.isRequired,
};

export default PolygonalListItem;
