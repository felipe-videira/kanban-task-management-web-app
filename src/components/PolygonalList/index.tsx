import { InferProps, number } from "prop-types";
import { useMemo } from "react";
import { PolygonalListContainer as Container } from "./styles";
import Item from "./PolygonalListItem";
import propTypes from "./propTypes";

function PolygonalList(props: InferProps<typeof PolygonalList.propTypes>) {
  const length = useMemo(() => props.data.length, [props.data]);
  const polygonExteriorAngle = useMemo(() => 360 / length, [length]);

  return length > 0 ? (
    <Container size={props.size} pointingUp={props.pointingUp}>
      <Item
        index={0}
        data={props.data}
        ItemComponent={props.ItemComponent}
        itemProps={props.itemProps}
        itemSize={props.itemSize}
        length={length}
        pointingUp={props.pointingUp}
        polygonExteriorAngle={polygonExteriorAngle}
      />
    </Container>
  ) : null;
}

PolygonalList.propTypes = {
  ...propTypes,
  size: number.isRequired,
};

PolygonalList.defaultProps = {
  pointingUp: true,
};

export default PolygonalList;
