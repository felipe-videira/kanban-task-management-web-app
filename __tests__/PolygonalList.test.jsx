import "jsdom-global/register";

import React from "react";
import { mount } from "enzyme";
import PolygonalList from "../src/components/PolygonalList";
import "jest-styled-components";

const mockProps = {
  data: [{ value: "value" }, { value: "value" }, { value: "value" }],
  ItemComponent: (props) => (
    // eslint-disable-next-line react/prop-types
    <button onClick={props.onClick}>{props.value}</button>
  ),
  itemProps: {
    onClick: () => {},
  },
  itemSize: 200,
  size: 400,
};

describe("PolygonalList", () => {
  it("should render correctly with all props", () => {
    const wrapper = mount(<PolygonalList {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should throw error if prop "data" is not present', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} data={undefined} />)
    ).toThrow();
  });

  it('should throw error if prop "data" is not an array of objects', () => {
    expect(() => shallow(<PolygonalList {...mockProps} data={1} />)).toThrow();
  });

  it('should throw error if prop "ItemComponent" is not present', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} ItemComponent={undefined} />)
    ).toThrow();
  });

  it('should throw error if prop "ItemComponent" is not a React element type', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} ItemComponent={() => {}} />)
    ).toThrow();
  });

  it('should throw error if prop "itemProps" is not an object', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} itemProps={"value"} />)
    ).toThrow();
  });

  it('should throw error if prop "itemSize" is not present', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} itemSize={undefined} />)
    ).toThrow();
  });

  it('should throw error if prop "itemSize" is not a number', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} itemSize={"1"} />)
    ).toThrow();
  });

  it('should throw error if prop "size" is not present', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} size={undefined} />)
    ).toThrow();
  });

  it('should throw error if prop "size" is not a number', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} size={"1"} />)
    ).toThrow();
  });
});
