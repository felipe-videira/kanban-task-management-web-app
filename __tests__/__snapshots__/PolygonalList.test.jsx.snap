// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolygonalList should render correctly with all props 1`] = `
.c0 {
  position: relative;
  overflow: hidden;
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
}

.c1::after {
  content: "";
  -webkit-transform-origin: -10px -10px;
  -ms-transform-origin: -10px -10px;
  transform-origin: -10px -10px;
  position: absolute;
  height: calc(50% / (3 / 4));
  width: 20px;
  background: rgba(0,0,0,0.2);
  -webkit-transform: rotate(calc((90deg - 120deg) / 2));
  -ms-transform: rotate(calc((90deg - 120deg) / 2));
  transform: rotate(calc((90deg - 120deg) / 2));
  top: calc(200px / 2);
  left: calc(200px / 2);
}

.c2 {
  position: absolute;
  top: 0;
  left: 0;
}

<div
  class="c0"
  size="400"
  style="width:400px;height:400px"
>
  <div
    class="c1"
    size="400"
    style="z-index:3;transform:rotate(45deg);top:calc(50% - 400px / 1.4142 / 2);left:calc(50% - 400px / 1.4142 / 2);width:calc(400px / 1.4142);height:calc(400px / 1.4142)"
  >
    <div
      class="c2"
      style="z-index:3;width:200px;height:200px;transform:rotate(calc(((120deg * 0) + 45deg) * -1))"
    />
    <div
      class="c1"
      size="400"
      style="z-index:2;transform:rotate(120deg);width:calc(400px / 1.4142);height:calc(400px / 1.4142)"
    >
      <div
        class="c2"
        style="z-index:2;width:200px;height:200px;transform:rotate(calc(((120deg * 1) + 45deg) * -1))"
      />
      <div
        class="c1"
        size="400"
        style="z-index:1;transform:rotate(120deg);width:calc(400px / 1.4142);height:calc(400px / 1.4142)"
      >
        <div
          class="c2"
          style="z-index:1;width:200px;height:200px;transform:rotate(calc(((120deg * 2) + 45deg) * -1))"
        />
      </div>
    </div>
  </div>
</div>
`;
